src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala

```diff
@@ -131,12 +132,17 @@ class AppsResource @Inject() (
     @PathParam("id") id: String,
     body: Array[Byte],
     @DefaultValue("false")@QueryParam("force") force: Boolean,
-    @Context req: HttpServletRequest, @Context resp: HttpServletResponse): Response = {
+    @Context req: HttpServletRequest): Response = doIfAuthenticated(req) { implicit identity =>
     val appId = id.toRootPath
-    withValid(Json.parse(body).as[AppUpdate].copy(id = Some(appId))) { app =>
-      doIfAuthorized(req, resp, UpdateAppOrGroup, appId) { identity =>
-        val now = clock.now()
-        val plan = result(groupManager.updateApp(appId, updateOrCreate(appId, _, app, now), now, force))
+    val now = clock.now()
+
+    withValid(Json.parse(body).as[AppUpdate].copy(id = Some(appId))) { appUpdate =>
+      val maybeExistingApp = result(groupManager.app(appId))
+      val updatedApp = updateOrCreate(appId, maybeExistingApp, appUpdate, now)
```

Comment:
groupManager.updateXXX makes each change transactional. If you load it outside the transaction, you can not rely on the value. Please move the authorization logic to `updateOrCreate`.